token-metadata/program/src/processor/verification/collection_verification.rs

```diff
@@ -0,0 +1,218 @@
+    assert_owned_by(collection_metadata_info, program_id)?;
+
+    let collection_master_edition_info = ctx
+        .accounts
+        .collection_master_edition_info
+        .ok_or(MetadataError::MissingCollectionMasterEdition)?;
+    assert_owned_by(collection_master_edition_info, program_id)?;
+
+    // Deserialize item metadata and collection parent metadata.
+    let mut metadata = Metadata::from_account_info(ctx.accounts.metadata_info)?;
+    let mut collection_metadata = Metadata::from_account_info(collection_metadata_info)?;
+
+    // Don't verify already verified items, otherwise for sized collections we end up with invalid
+    // size data.
+    if let Some(collection) = &metadata.collection {
+        if collection.verified {
+            return Err(MetadataError::AlreadyVerified.into());
+        }
+    }
+
+    // Verify the collection in the item's metadata matches the collection mint.  Also verify
+    // the collection metadata matches the collection mint, and the collection edition derivation.
+    assert_collection_verify_is_valid(
+        &metadata.collection,
+        &collection_metadata,
+        collection_mint_info,
+        collection_master_edition_info,
+    )?;
+
+    // Determines if we have a valid authority to perform the collection verification.  The
+    // required authority is either the collection parent's metadata update authority, or a
+    // collection delegate for the collection parent.  This call fails if no valid authority is
+    // present.
+    let authority_response = AuthorityType::get_authority_type(AuthorityRequest {
+        authority: ctx.accounts.authority_info.key,
+        update_authority: &collection_metadata.update_authority,
+        mint: collection_mint_info.key,
+        metadata_delegate_record_info: ctx.accounts.delegate_record_info,
+        metadata_delegate_roles: vec![MetadataDelegateRole::Collection],
+        precedence: &[AuthorityType::Metadata, AuthorityType::MetadataDelegate],
+        ..Default::default()
+    })?;
+
+    // Validate that authority type is expected.
+    match authority_response.authority_type {
+        AuthorityType::Metadata | AuthorityType::MetadataDelegate => (),
+        _ => return Err(MetadataError::UpdateAuthorityIncorrect.into()),
+    }
+
+    // Destructure the collection field from the item metadata.
+    match metadata.collection.as_mut() {
+        Some(collection) => {
+            // Set item metadata collection to verified.
+            collection.verified = true;
+
+            // In the case of a sized collection, update the size on the collection parent.
+            if collection_metadata.collection_details.is_some() {
+                increment_collection_size(&mut collection_metadata, collection_metadata_info)?;
+            }
+        }
+        None => return Err(MetadataError::CollectionNotFound.into()),
+    };
+
+    // Reserialize metadata.
+    clean_write_metadata(&mut metadata, ctx.accounts.metadata_info)?;
```

Comment:
Once stylistic suggestion here is you can elide the `Ok(())` and return the result from the `clean_write_metadata` call directly:

```rust
     . . .
    clean_write_metadata(&mut metadata, ctx.accounts.metadata_info)
}
```